include/pybind11/detail/internals.h

```diff
@@ -421,6 +421,39 @@ inline void translate_local_exception(std::exception_ptr p) {
 }
 #endif
 
+inline object get_python_state_dict() {
+    object state_dict;
+#if (PYBIND11_INTERNALS_VERSION <= 4 && PY_VERSION_HEX < 0x030C0000)                              \
```

Comment:
I think it might be simpler if you just had a define that you are always using internals version 5 if the python version is 3.12

Reply:
I added this commit: a version bump for Python 3.12+ (I think of it as a free lunch version bump):

```diff
 #ifndef PYBIND11_INTERNALS_VERSION
-#    define PYBIND11_INTERNALS_VERSION 4
+#    if PY_VERSION_HEX >= 0x030C0000
+// Version bump for Python 3.12+, before first 3.12 beta release.
+#        define PYBIND11_INTERNALS_VERSION 5
+#    else
+#        define PYBIND11_INTERNALS_VERSION 4
+#    endif
 #endif
```

Then I looked long and hard at the condition here but eventually decided to leave it at a no-op shuffle, solely to increase readability slightly (see e8b34b473546c469f11cd2c2a0cb098ca12134e5).

> it might be simpler

Did you have in mind removing the `PY_VERSION_HEX < 0x030C0000` sub-condition here?

What worries me: if someone defines `PYBIND11_INTERNALS_VERSION 4` externally, we'd select the wrong code here if the sub-condition is missing. We don't know yet what the resulting errors might be, if any, for future Python releases.

I guess we could do something like this to guard the `PyEval_GetBuiltins()` path:

```
#if PY_VERSION_HEX >= 0x030C0000
#error Python 3.12+ requires PYBIND11_INTERNALS_VERSION >= 5
#endif
```

But that's strictly speaking an arbitrary forcing function. Maybe someone will have a good reason to experiment with `PYBIND11_INTERNALS_VERSION` 4 even for 3.12+? E.g. when debugging some issue with the TLS-related changes connected to version 5. This is totally made up of course, but I'm generally trying to stay clear of non-essential forcing functions (anecdotes of getting burned omitted).

Could you please re-approve if you think the latest version here is fine, or let me know if I misunderstood what you had in mind?

BTW: I re-tested with Python 3.12.0a6 under google/pywrapcc#30019 (https://github.com/google/pywrapcc/actions/runs/4526056859/jobs/7970965421?pr=30019).

Reply:
> But that's strictly speaking an arbitrary forcing function.

Yeah, it's an "arbitrary forcing function", but it reduces the matrix of Python version / pybind11 version abi combinations you have to support. I don't understand why you would want version abi version 4 and python 3.12 together, but if you do, feel free to disregard this suggestion because I was assuming that combination was sorta worthless (since you can't get abi version 4 and python 3.11 code to work together anyways).

Your code in mind is exactly what I had in mind, but here are a couple of suggestions.

```
#if PY_VERSION_HEX < 0x03080000                                                                   \
    || (PYBIND11_INTERNALS_VERSION <= 4 && PY_VERSION_HEX < 0x030C0000) || defined(PYPY_VERSION)
    state_dict = reinterpret_borrow<object>(PyEval_GetBuiltins());
#else
```

Why isn't this just

```
#if PYBIND11_INTERNALS_VERSION <= 4
    state_dict = reinterpret_borrow<object>(PyEval_GetBuiltins());
#else
```

> I guess we could do something like this to guard the PyEval_GetBuiltins() path:

I would add this to the very top where versions are defined
```

#ifndef PYBIND11_INTERNALS_VERSION
#    if PY_VERSION_HEX >= 0x030C0000
// Version bump for Python 3.12+, before first 3.12 beta release.
#        define PYBIND11_INTERNALS_VERSION 5
#    else
#        define PYBIND11_INTERNALS_VERSION 4
#    endif
#endif

// Require at least version 5 ABI for Python 3.12+
static_assert(!(PY_VERSION_HEX == 0x030C0000) || (PYBIND11_INTERNALS_VERSION >= 5);

```

Reply:
> Why isn't this just
> 
> ```
> #if PYBIND11_INTERNALS_VERSION <= 4
> ```

We need to keep the < Python 3.8 condition because `_PyInterpreterState_Get()` only appeared with Python 3.8 (I looked in the 3.6 and 3.7 sources yesterday).

I'm trying the simpler condition right now. I'm not sure what we need to do for PyPy.

I'm on board enforcing ABI version 5 for 3.12+ (see latest code here), but just to make sure I'm not missing something:

> (since you can't get abi version 4 and python 3.11 code to work together anyways)

Did you mean to write 3.12 here?  —  We are testing with 3.11, e.g.:

```
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-7.0.0, pluggy-1.0.0
C++ Info: 9.4.0 C++11 __pybind11_internals_v4_gcc_libstdcpp_cxxabi1013__ PYBIND11_SIMPLE_GIL_MANAGEMENT=True
```

Currently ABI 4 & 3.12 still work (from [pywrapcc GHA](https://github.com/google/pywrapcc/actions/runs/4461749631/jobs/7835824105)):

```
============================= test session starts ==============================
platform linux -- Python 3.12.0a6, pytest-7.0.0, pluggy-1.0.0
C++ Info: 11.3.0 C++11 __pybind11_internals_v4_gcc_libstdcpp_cxxabi1016_sh_def__ PYBIND11_SIMPLE_GIL_MANAGEMENT=False
```

Reply:
> (since you can't get abi version 4 and python 3.11 code to work together anyways)

Oops, typo. Yep, Meant that you can't get abi version 4 Python 3.12 code and abi version 4 Python 3.11 code to work together

> I'm not sure what we need to do for PyPy.

PyPy should be following the CPython API, so we should be able to use identical code for both it and CPython. Unless you know otherwise of course.

> We need to keep the < Python 3.8 condition because _PyInterpreterState_Get() only appeared with Python 3.8 (I looked in the 3.6 and 3.7 sources yesterday).

I would add a wrapper for _PyInterpreterState_Get to pybind11/detail/common.h with #ifdefs to switch back and forth between the _ and regular version depending on the Python version.

Reply:
> PyPy should be following the CPython API, so we should be able to use identical code for both it and CPython.

Unfortunately not. PyPy with 3.8 and 3.9 still don't work (our GHA). I just looked in the [latest PyPy sources](https://foss.heptapod.net/pypy/pypy): `PyInterpreterState_Get` still doesn't appear there. I tried to use `PY_VERSION_HEX` but apparently that's not meaningful with PyPy. It's unclear to me at the moment if/how we can do better than just testing for `PYPY_VERSION`. Leaving that for later. While looking around I saw that there are 7.3.10 and 7.3.11 releases which we don't have in our GHA. Something to do separately maybe.

> I would add a wrapper for _PyInterpreterState_Get to pybind11/detail/common.h with #ifdefs to switch back and forth between the _ and regular version depending on the Python version.

This `get_python_state_dict()` is essentially that already. I don't feel factoring out the branch for >= 3.9 will be helpful, especially given the PyPy complication.

Reply:
Arg, another unfortunate PyPy incompatibility. I guess this is the best we can do.